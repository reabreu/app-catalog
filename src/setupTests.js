// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";
import { rest } from "msw";
import { setupServer } from "msw/node";

const handlers = [
  rest.get("/apps", (req, res, ctx) => {
    return res(
      ctx.json([
        {
          id: "asdm123312asdnboqwedsa",
          name: "prometheus-operator-app-chart",
          description: "A Helm Chart for Prometheus Operator.",
          version: "0.2.0",
          author: "Giantswarm",
          iconURL:
            "https://raw.githubusercontent.com/prometheus/prometheus.github.io/master/assets/prometheus_logo-cb55bb5c346.png",
          url: "https://github.com/giantswarm/prometheus-operator-app",
          readmeURL:
            "https://raw.githubusercontent.com/giantswarm/prometheus-operator-app/v0.1.1/README.md",
        },
        {
          id: "dsjklnsankj12831209asmdo",
          name: "nginx-ingress-controller-app",
          description: "A Helm chart for the nginx ingress-controller.",
          version: "1.6.9",
          author: "Giantswarm",
          iconURL:
            "https://upload.wikimedia.org/wikipedia/commons/thumb/c/c5/Nginx_logo.svg/500px-Nginx_logo.svg.png",
          url: "https://github.com/giantswarm/nginx-ingress-controller-app",
          readmeURL:
            "https://raw.githubusercontent.com/giantswarm/nginx-ingress-controller-app/v1.6.9/README.md",
        },
      ])
    );
  }),
];

export const server = setupServer(...handlers);

beforeAll(() => server.listen());
// if you need to add a handler after calling setupServer for some specific test
// this will remove that handler for the rest of them
// (which is important for test isolation):
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
