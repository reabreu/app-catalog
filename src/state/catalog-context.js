import React, { useReducer } from "react";

const CatalogStateContext = React.createContext();
const CatalogDispatchContext = React.createContext();

export const ACTIONS = Object.freeze({
  ADD: "ADD",
  ADD_BY_ID: "ADD_BY_ID",
  SETSTATUS: "SETSTATUS",
});

export const STATUS = Object.freeze({
  IDLE: "IDLE",
  FETCHING: "FETCHING",
  ERROR: "ERROR",
});

// Reducer
function catalogReducer(state, action) {
  switch (action.type) {
    case ACTIONS.ADD: {
      const newApps = {};

      action.payload.forEach((element) => {
        newApps[element.id] = element;
      });

      return { ...state, apps: newApps };
    }
    case ACTIONS.SETSTATUS: {
      return { ...state, status: action.payload };
    }
    case ACTIONS.ADD_BY_ID: {
      return {
        ...state,
        apps: {
          ...state.apps,
          [action.payload.id]: action.payload.data,
        },
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
}

// Async actions
async function fetchAllApps(dispatch) {
  dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.FETCHING });

  try {
    const request = await fetch("/apps");
    const res = await request.json();

    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.IDLE });
    dispatch({ type: ACTIONS.ADD, payload: res });
  } catch (error) {
    console.log("error", error);
    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.ERROR });
  }
}

async function fetchAppById(dispatch, id) {
  dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.FETCHING });

  try {
    const request = await fetch(`/apps/${id}`);
    const res = await request.json();

    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.IDLE });
    dispatch({
      type: ACTIONS.ADD_BY_ID,
      payload: {
        id,
        data: res,
      },
    });
  } catch (error) {
    console.log("error", error);
    dispatch({ type: ACTIONS.SETSTATUS, payload: STATUS.ERROR });
  }
}

const initialState = {
  apps: {},
  status: STATUS.IDLE,
};

function CatalogProvider({ children }) {
  const [state, dispatch] = useReducer(catalogReducer, initialState);

  return (
    <CatalogStateContext.Provider value={state}>
      <CatalogDispatchContext.Provider value={dispatch}>
        {children}
      </CatalogDispatchContext.Provider>
    </CatalogStateContext.Provider>
  );
}

function useCatalogState() {
  const context = React.useContext(CatalogStateContext);
  if (context === undefined) {
    throw new Error("useCountState must be used within a CatalogProvider");
  }
  return context;
}

function useCatalogDispatch() {
  const context = React.useContext(CatalogDispatchContext);
  if (context === undefined) {
    throw new Error("useCountDispatch must be used within a CatalogProvider");
  }
  return context;
}

function useCatalogContext() {
  return [useCatalogState(), useCatalogDispatch()];
}

export {
  CatalogProvider,
  useCatalogContext,
  useCatalogState,
  useCatalogDispatch,
  fetchAllApps,
  fetchAppById,
};
