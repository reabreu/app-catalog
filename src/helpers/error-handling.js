import React from "react";
import { useCatalogState, STATUS } from "../state/catalog-context";
import { FetchingText, RetryButton } from "./styles";

export const RenderWithErrorHandling = ({ children, onErrorRetry }) => {
  const { status } = useCatalogState();

  const isFetching = STATUS.FETCHING === status;
  const isError = STATUS.ERROR === status;
  const isIdle = STATUS.IDLE === status;

  return (
    <>
      {isIdle && children}
      {isFetching && (
        <FetchingText>
          Hang in there a little bit, we're fetching the app for you...
        </FetchingText>
      )}
      {isError && (
        <RetryButton onClick={onErrorRetry}>
          Oops, something went wrong, click here to retry
        </RetryButton>
      )}
    </>
  );
};
