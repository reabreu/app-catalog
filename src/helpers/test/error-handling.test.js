import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { rest } from "msw";
import * as CatalogContext from "../../state/catalog-context";
import { server } from "../../setupTests";
import { RenderWithErrorHandling } from "../error-handling";

describe("<RenderWithErrorHandling /> spec", () => {
  beforeEach(() => {
    jest.spyOn(CatalogContext, "useCatalogState");
  });

  it("Renders children on IDLE", () => {
    CatalogContext.useCatalogState.mockImplementation(() => ({
      status: CatalogContext.STATUS.IDLE,
    }));

    const { asFragment } = render(
      <CatalogContext.CatalogProvider>
        <RenderWithErrorHandling>
          <p>children</p>
        </RenderWithErrorHandling>
      </CatalogContext.CatalogProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders fetching text on FETCHING", () => {
    CatalogContext.useCatalogState.mockImplementation(() => ({
      status: CatalogContext.STATUS.FETCHING,
    }));

    const { asFragment } = render(
      <CatalogContext.CatalogProvider>
        <RenderWithErrorHandling>
          <p>children</p>
        </RenderWithErrorHandling>
      </CatalogContext.CatalogProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders error button text on ERROR", async () => {
    CatalogContext.useCatalogState.mockImplementation(() => ({
      status: CatalogContext.STATUS.ERROR,
    }));

    const myMock = jest.fn();

    const { asFragment, findByRole } = render(
      <CatalogContext.CatalogProvider>
        <RenderWithErrorHandling onErrorRetry={myMock}>
          <p>children</p>
        </RenderWithErrorHandling>
      </CatalogContext.CatalogProvider>
    );

    expect(asFragment()).toMatchSnapshot();

    const button = await findByRole("button");

    fireEvent.click(button);

    expect(myMock).toHaveBeenCalled();
  });
});
