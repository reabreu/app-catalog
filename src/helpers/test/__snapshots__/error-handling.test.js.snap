// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RenderWithErrorHandling /> spec Renders children on IDLE 1`] = `
<DocumentFragment>
  <p>
    children
  </p>
</DocumentFragment>
`;

exports[`<RenderWithErrorHandling /> spec Renders error button text on ERROR 1`] = `
<DocumentFragment>
  <button
    class="sc-AxjAm hUNfZb"
  >
    Oops, something went wrong, click here to retry
  </button>
</DocumentFragment>
`;

exports[`<RenderWithErrorHandling /> spec Renders fetching text on FETCHING 1`] = `
<DocumentFragment>
  <p
    class="sc-AxirZ kveKhk"
  >
    Hang in there a little bit, we're fetching the app for you...
  </p>
</DocumentFragment>
`;
